frame/transaction-payment/src/lib.rs

```diff
@@ -207,28 +207,36 @@ impl<T: Trait + Send + Sync> SignedExtension for ChargeTransactionPayment<T>
 		info: Self::DispatchInfo,
 		len: usize,
 	) -> TransactionValidity {
-		// pay any fees.
+		// pay any fees and tip.
 		let tip = self.0;
-		let fee = Self::compute_fee(len as u32, info, tip);
-		let imbalance = match T::Currency::withdraw(
+		let tip_imbalance = match T::Currency::withdraw(
+			who,
+			tip,
+			WithdrawReason::Tip.into(),
+			ExistenceRequirement::KeepAlive,
+		) {
+			Ok(imbalance) => imbalance,
+			Err(_) => return InvalidTransaction::Payment.into(),
+		};
+
+		let fee = Self::compute_fee(len as u32, info);
+		let fee_imbalance = match T::Currency::withdraw(
 			who,
 			fee,
-			if tip.is_zero() {
-				WithdrawReason::TransactionPayment.into()
-			} else {
-				WithdrawReason::TransactionPayment | WithdrawReason::Tip
-			},
+			WithdrawReason::TransactionPayment.into(),
 			ExistenceRequirement::KeepAlive,
 		) {
 			Ok(imbalance) => imbalance,
 			Err(_) => return InvalidTransaction::Payment.into(),
 		};
-		T::OnTransactionPayment::on_unbalanced(imbalance);
+
+		T::OnTransactionTipPayment::on_unbalanced(tip_imbalance);
+		T::OnTransactionFeePayment::on_unbalanced(fee_imbalance);
 
 		let mut r = ValidTransaction::default();
 		// NOTE: we probably want to maximize the _fee (of any type) per weight unit_ here, which
 		// will be a bit more than setting the priority to tip. For now, this is enough.
-		r.priority = fee.saturated_into::<TransactionPriority>();
+		r.priority = fee.saturating_add(tip).saturated_into::<TransactionPriority>();
```

Comment:
This priority is now biased as the tip+fee is not actually proportional to the amount received by validator